<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1">
<title>SnowTools: GenomicRegionVector Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.4.7 -->
<div class="tabs">
  <ul>
    <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
    <li id="current"><a href="annotated.html"><span>Classes</span></a></li>
    <li><a href="files.html"><span>Files</span></a></li>
  </ul></div>
<div class="tabs">
  <ul>
    <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
    <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
  </ul></div>
<h1>GenomicRegionVector Class Reference</h1><!-- doxytag: class="GenomicRegionVector" --><code>#include &lt;<a class="el" href="GenomicRegionVector_8h-source.html">GenomicRegionVector.h</a>&gt;</code>
<p>
<a href="classGenomicRegionVector-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#220b3dce8817e647aa454f573b6306d4">GenomicRegionVector</a> ()</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#818072b55ff52c612f2f32952d2ebfc4">GenomicRegionVector</a> (std::vector&lt; GenomicRegion &gt; &amp;vec)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#9c3771802a2e428fbd26b56083de5da2">GenomicRegionVector</a> (int width, int ovlp, const GenomicRegion &amp;gr)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#e4a4d90b656a928773581de39a26c580">readMuTect</a> (const std::string &amp;file, int pad=0)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#ba5589c933030ffb64dbeaf8350276b2">readBEDfile</a> (const std::string &amp;file, int pad=0)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#27b8fa36a7b5b81f5abe2dad8dd752a7">readVCFfile</a> (const std::string &amp;file, int pad=0)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#e3e69a79c878ae9d4f62b74e5f08bc5b">regionFileToGRV</a> (const std::string &amp;file, int pad=0)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#45fd5c44e0851148b17146318795404d">createTreeMap</a> ()</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#d6c335703ad494ce93e0e87e8bf9e493">sendToBED</a> (const std::string file)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#a30d9f4def4c7ac7958a4475f2997df5">mergeOverlappingIntervals</a> ()</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">size_t&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#7eefd44e17d6a72a252ce733381982b3">size</a> () const </td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#84015cecf9ef013ca6a29018dc83b8c4">add</a> (const GenomicRegion &amp;g)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#aea0228989aeb752d857b213eeaf8e2d">empty</a> () const </td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#5d86f830a20b787b13ea682c1bb4e523">clear</a> ()</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">const GenomicRegion &amp;&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#3b80a5a66948d30c5918a9878a8ac392">at</a> (size_t i) const </td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">size_t&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#b9895bc9c23c687561b3c23ac0f23320">findOverlapping</a> (const GenomicRegion &amp;gr)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#894ce6c8a5b366f5fb3f75a310d2499e">concat</a> (const <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a> &amp;g)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#2fddfd8ce536ef6ea82f221a1fbaa9a1">sendToBED</a> () const </td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#9880c8cdbe9ac50d8a579b483d0649f5">getNextGenomicRegion</a> (GenomicRegion &amp;gr)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classGenomicRegionVector.html#134e5edbdb66be9149dcc1bae97c4b7a">rewind</a> ()</td></tr>

</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
Class to store vector of intervals on the genome 
<p>
<hr><h2>Constructor &amp; Destructor Documentation</h2>
<a class="anchor" name="220b3dce8817e647aa454f573b6306d4"></a><!-- doxytag: member="GenomicRegionVector::GenomicRegionVector" ref="220b3dce8817e647aa454f573b6306d4" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">GenomicRegionVector::GenomicRegionVector           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"><code> [inline]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Construct an empty <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a> 
</div>
</div><p>
<a class="anchor" name="818072b55ff52c612f2f32952d2ebfc4"></a><!-- doxytag: member="GenomicRegionVector::GenomicRegionVector" ref="818072b55ff52c612f2f32952d2ebfc4" args="(std::vector&lt; GenomicRegion &gt; &amp;vec)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">GenomicRegionVector::GenomicRegionVector           </td>
          <td>(</td>
          <td class="paramtype">std::vector&lt; GenomicRegion &gt; &amp;&nbsp;</td>
          <td class="paramname"> <em>vec</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Construct from a plain vector of GenomicRegion objects 
</div>
</div><p>
<a class="anchor" name="9c3771802a2e428fbd26b56083de5da2"></a><!-- doxytag: member="GenomicRegionVector::GenomicRegionVector" ref="9c3771802a2e428fbd26b56083de5da2" args="(int width, int ovlp, const GenomicRegion &amp;gr)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">GenomicRegionVector::GenomicRegionVector           </td>
          <td>(</td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>width</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>ovlp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const GenomicRegion &amp;&nbsp;</td>
          <td class="paramname"> <em>gr</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Construct a <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a> with overlapping intervals<p>
<dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>width</em>&nbsp;</td><td>Desired bin width </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>ovlp</em>&nbsp;</td><td>Amount that the bins should overlap </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>gr</em>&nbsp;</td><td>GenomicRegion to divide into smaller overlapping bins </td></tr>
  </table>
</dl>

</div>
</div><p>
<hr><h2>Member Function Documentation</h2>
<a class="anchor" name="84015cecf9ef013ca6a29018dc83b8c4"></a><!-- doxytag: member="GenomicRegionVector::add" ref="84015cecf9ef013ca6a29018dc83b8c4" args="(const GenomicRegion &amp;g)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionVector::add           </td>
          <td>(</td>
          <td class="paramtype">const GenomicRegion &amp;&nbsp;</td>
          <td class="paramname"> <em>g</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"><code> [inline]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Add a new GenomicRegion to end 
</div>
</div><p>
<a class="anchor" name="3b80a5a66948d30c5918a9878a8ac392"></a><!-- doxytag: member="GenomicRegionVector::at" ref="3b80a5a66948d30c5918a9878a8ac392" args="(size_t i) const " -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">const GenomicRegion&amp; GenomicRegionVector::at           </td>
          <td>(</td>
          <td class="paramtype">size_t&nbsp;</td>
          <td class="paramname"> <em>i</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"> const</td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Retrieve a GenomicRegion at given index.<p>
Note that this does not move the idx iterator, which is used to loop through all the regions. Throws an exception if the index is out of bounds. <dl compact><dt><b>Returns:</b></dt><dd>GenomicRegion pointed to by index i </dd></dl>

</div>
</div><p>
<a class="anchor" name="5d86f830a20b787b13ea682c1bb4e523"></a><!-- doxytag: member="GenomicRegionVector::clear" ref="5d86f830a20b787b13ea682c1bb4e523" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionVector::clear           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"><code> [inline]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Clear out all of the GenomicRegion objects 
</div>
</div><p>
<a class="anchor" name="894ce6c8a5b366f5fb3f75a310d2499e"></a><!-- doxytag: member="GenomicRegionVector::concat" ref="894ce6c8a5b366f5fb3f75a310d2499e" args="(const GenomicRegionVector &amp;g)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionVector::concat           </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>g</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Add two <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a> objects together 
</div>
</div><p>
<a class="anchor" name="45fd5c44e0851148b17146318795404d"></a><!-- doxytag: member="GenomicRegionVector::createTreeMap" ref="45fd5c44e0851148b17146318795404d" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionVector::createTreeMap           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Fill in the GenomicIntervalTreeMap stored in this object.<p>
A GenomicIntervalTreeMap is an unordered_map of GenomicIntervalTrees for each chromosome. A GenomicIntervalTree is an interval tree on the ranges defined by the genomic interval, with cargo set at the same GenomicRegion object. 
</div>
</div><p>
<a class="anchor" name="aea0228989aeb752d857b213eeaf8e2d"></a><!-- doxytag: member="GenomicRegionVector::empty" ref="aea0228989aeb752d857b213eeaf8e2d" args="() const " -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool GenomicRegionVector::empty           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"> const<code> [inline]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Is this object empty? 
</div>
</div><p>
<a class="anchor" name="b9895bc9c23c687561b3c23ac0f23320"></a><!-- doxytag: member="GenomicRegionVector::findOverlapping" ref="b9895bc9c23c687561b3c23ac0f23320" args="(const GenomicRegion &amp;gr)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t GenomicRegionVector::findOverlapping           </td>
          <td>(</td>
          <td class="paramtype">const GenomicRegion &amp;&nbsp;</td>
          <td class="paramname"> <em>gr</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Find overlaps between this vector and input GenomicRegion.<p>
Requires that the GenomicIntervalTreeMap have been created first <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>gr</em>&nbsp;</td><td>Region to test </td></tr>
  </table>
</dl>
<dl compact><dt><b>Returns:</b></dt><dd>Number of overlapping elements in this <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a> </dd></dl>

</div>
</div><p>
<a class="anchor" name="9880c8cdbe9ac50d8a579b483d0649f5"></a><!-- doxytag: member="GenomicRegionVector::getNextGenomicRegion" ref="9880c8cdbe9ac50d8a579b483d0649f5" args="(GenomicRegion &amp;gr)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool GenomicRegionVector::getNextGenomicRegion           </td>
          <td>(</td>
          <td class="paramtype">GenomicRegion &amp;&nbsp;</td>
          <td class="paramname"> <em>gr</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Fill the next GenomicRegion object. <dl compact><dt><b>Returns:</b></dt><dd>false if add end of vector </dd></dl>

</div>
</div><p>
<a class="anchor" name="a30d9f4def4c7ac7958a4475f2997df5"></a><!-- doxytag: member="GenomicRegionVector::mergeOverlappingIntervals" ref="a30d9f4def4c7ac7958a4475f2997df5" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionVector::mergeOverlappingIntervals           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Reduces the GenomicRegion objects to minimal set 
</div>
</div><p>
<a class="anchor" name="ba5589c933030ffb64dbeaf8350276b2"></a><!-- doxytag: member="GenomicRegionVector::readBEDfile" ref="ba5589c933030ffb64dbeaf8350276b2" args="(const std::string &amp;file, int pad=0)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionVector::readBEDfile           </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&nbsp;</td>
          <td class="paramname"> <em>file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>pad</em> = <code>0</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Read in a BED file and adds to <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a> object <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>file</em>&nbsp;</td><td>Path to BED file </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>pad</em>&nbsp;</td><td>Amount to pad intervals by </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="e4a4d90b656a928773581de39a26c580"></a><!-- doxytag: member="GenomicRegionVector::readMuTect" ref="e4a4d90b656a928773581de39a26c580" args="(const std::string &amp;file, int pad=0)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionVector::readMuTect           </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&nbsp;</td>
          <td class="paramname"> <em>file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>pad</em> = <code>0</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Read in a MuTect call-stats file and adds to <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a> object.<p>
Reads a MuTect call-stats file and imports only lines with KEEP marked. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>file</em>&nbsp;</td><td>Path to call-stats file </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>pad</em>&nbsp;</td><td>Amount to pad intervals by </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="27b8fa36a7b5b81f5abe2dad8dd752a7"></a><!-- doxytag: member="GenomicRegionVector::readVCFfile" ref="27b8fa36a7b5b81f5abe2dad8dd752a7" args="(const std::string &amp;file, int pad=0)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionVector::readVCFfile           </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&nbsp;</td>
          <td class="paramname"> <em>file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>pad</em> = <code>0</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Read in a VCF file and adds to <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a> object <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>file</em>&nbsp;</td><td>Path to BED file </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>pad</em>&nbsp;</td><td>Amount to pad intervals by </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="e3e69a79c878ae9d4f62b74e5f08bc5b"></a><!-- doxytag: member="GenomicRegionVector::regionFileToGRV" ref="e3e69a79c878ae9d4f62b74e5f08bc5b" args="(const std::string &amp;file, int pad=0)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionVector::regionFileToGRV           </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&nbsp;</td>
          <td class="paramname"> <em>file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&nbsp;</td>
          <td class="paramname"> <em>pad</em> = <code>0</code></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Read in a text file (can be gzipped) and add to <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a><p>
This function will automatically detect which file type is being input: -- ends in .vcf -&gt; readVCFfile -- ends in .bed -&gt; readBEDfile -- header contains "MuTect" -&gt; readMuTect The values are appended to existing vector of GenomicRegion objects <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>file</em>&nbsp;</td><td>Text file to read and store intervals </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>pad</em>&nbsp;</td><td>Amount to pad the intervals by (calls GenomicRegion::pad) </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="134e5edbdb66be9149dcc1bae97c4b7a"></a><!-- doxytag: member="GenomicRegionVector::rewind" ref="134e5edbdb66be9149dcc1bae97c4b7a" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionVector::rewind           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"><code> [inline]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Rewind the element pointer to the first GenomicRegion 
</div>
</div><p>
<a class="anchor" name="2fddfd8ce536ef6ea82f221a1fbaa9a1"></a><!-- doxytag: member="GenomicRegionVector::sendToBED" ref="2fddfd8ce536ef6ea82f221a1fbaa9a1" args="() const " -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">string GenomicRegionVector::sendToBED           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"> const</td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Output the <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a> to a BED format<p>
Currently just outputs chr pos pos2 with no header. <dl compact><dt><b>Returns:</b></dt><dd>BED formated string reprsentation </dd></dl>

</div>
</div><p>
<a class="anchor" name="d6c335703ad494ce93e0e87e8bf9e493"></a><!-- doxytag: member="GenomicRegionVector::sendToBED" ref="d6c335703ad494ce93e0e87e8bf9e493" args="(const std::string file)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void GenomicRegionVector::sendToBED           </td>
          <td>(</td>
          <td class="paramtype">const std::string&nbsp;</td>
          <td class="paramname"> <em>file</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Send the <a class="el" href="classGenomicRegionVector.html">GenomicRegionVector</a> to a BED file <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>file</em>&nbsp;</td><td>Output file </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="7eefd44e17d6a72a252ce733381982b3"></a><!-- doxytag: member="GenomicRegionVector::size" ref="7eefd44e17d6a72a252ce733381982b3" args="() const " -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t GenomicRegionVector::size           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"> const<code> [inline]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Return the number of GenomicRegions stored 
</div>
</div><p>
<hr>The documentation for this class was generated from the following file:<ul>
<li>/xchip/gistic/Jeremiah/GIT/isva/SnowTools/src/SnowTools/<a class="el" href="GenomicRegionVector_8h-source.html">GenomicRegionVector.h</a></ul>
<hr size="1"><address style="align: right;"><small>Generated on Fri Apr 24 15:50:15 2015 for SnowTools by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.4.7 </small></address>
</body>
</html>
