\section{Genomic\-Region\-Vector Class Reference}
\label{classGenomicRegionVector}\index{GenomicRegionVector@{GenomicRegionVector}}
{\tt \#include $<$Genomic\-Region\-Vector.h$>$}

\subsection*{Public Member Functions}
\begin{CompactItemize}
\item 
\bf{Genomic\-Region\-Vector} ()
\item 
\bf{Genomic\-Region\-Vector} (std::vector$<$ Genomic\-Region $>$ \&vec)
\item 
\bf{Genomic\-Region\-Vector} (int width, int ovlp, const Genomic\-Region \&gr)
\item 
void \bf{read\-Mu\-Tect} (const std::string \&file, int pad=0)
\item 
void \bf{read\-BEDfile} (const std::string \&file, int pad=0)
\item 
void \bf{read\-VCFfile} (const std::string \&file, int pad=0)
\item 
void \bf{region\-File\-To\-GRV} (const std::string \&file, int pad=0)
\item 
void \bf{create\-Tree\-Map} ()
\item 
void \bf{send\-To\-BED} (const std::string file)
\item 
void \bf{merge\-Overlapping\-Intervals} ()
\item 
size\_\-t \bf{size} () const 
\item 
void \bf{add} (const Genomic\-Region \&g)
\item 
bool \bf{empty} () const 
\item 
void \bf{clear} ()
\item 
const Genomic\-Region \& \bf{at} (size\_\-t i) const 
\item 
size\_\-t \bf{find\-Overlapping} (const Genomic\-Region \&gr)
\item 
void \bf{concat} (const \bf{Genomic\-Region\-Vector} \&g)
\item 
string \bf{send\-To\-BED} () const 
\item 
bool \bf{get\-Next\-Genomic\-Region} (Genomic\-Region \&gr)
\item 
void \bf{rewind} ()
\end{CompactItemize}


\subsection{Detailed Description}
Class to store vector of intervals on the genome 



\subsection{Constructor \& Destructor Documentation}
\index{GenomicRegionVector@{Genomic\-Region\-Vector}!GenomicRegionVector@{GenomicRegionVector}}
\index{GenomicRegionVector@{GenomicRegionVector}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}Genomic\-Region\-Vector::Genomic\-Region\-Vector ()\hspace{0.3cm}{\tt  [inline]}}\label{classGenomicRegionVector_220b3dce8817e647aa454f573b6306d4}


Construct an empty \doxyref{Genomic\-Region\-Vector}{p.}{classGenomicRegionVector} \index{GenomicRegionVector@{Genomic\-Region\-Vector}!GenomicRegionVector@{GenomicRegionVector}}
\index{GenomicRegionVector@{GenomicRegionVector}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}Genomic\-Region\-Vector::Genomic\-Region\-Vector (std::vector$<$ Genomic\-Region $>$ \& {\em vec})}\label{classGenomicRegionVector_818072b55ff52c612f2f32952d2ebfc4}


Construct from a plain vector of Genomic\-Region objects \index{GenomicRegionVector@{Genomic\-Region\-Vector}!GenomicRegionVector@{GenomicRegionVector}}
\index{GenomicRegionVector@{GenomicRegionVector}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}Genomic\-Region\-Vector::Genomic\-Region\-Vector (int {\em width}, int {\em ovlp}, const Genomic\-Region \& {\em gr})}\label{classGenomicRegionVector_9c3771802a2e428fbd26b56083de5da2}


Construct a \doxyref{Genomic\-Region\-Vector}{p.}{classGenomicRegionVector} with overlapping intervals

\begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em width}]Desired bin width \item[{\em ovlp}]Amount that the bins should overlap \item[{\em gr}]Genomic\-Region to divide into smaller overlapping bins \end{description}
\end{Desc}


\subsection{Member Function Documentation}
\index{GenomicRegionVector@{Genomic\-Region\-Vector}!add@{add}}
\index{add@{add}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Genomic\-Region\-Vector::add (const Genomic\-Region \& {\em g})\hspace{0.3cm}{\tt  [inline]}}\label{classGenomicRegionVector_84015cecf9ef013ca6a29018dc83b8c4}


Add a new Genomic\-Region to end \index{GenomicRegionVector@{Genomic\-Region\-Vector}!at@{at}}
\index{at@{at}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}const Genomic\-Region\& Genomic\-Region\-Vector::at (size\_\-t {\em i}) const}\label{classGenomicRegionVector_3b80a5a66948d30c5918a9878a8ac392}


Retrieve a Genomic\-Region at given index.

Note that this does not move the idx iterator, which is used to loop through all the regions. Throws an exception if the index is out of bounds. \begin{Desc}
\item[Returns:]Genomic\-Region pointed to by index i \end{Desc}
\index{GenomicRegionVector@{Genomic\-Region\-Vector}!clear@{clear}}
\index{clear@{clear}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Genomic\-Region\-Vector::clear ()\hspace{0.3cm}{\tt  [inline]}}\label{classGenomicRegionVector_5d86f830a20b787b13ea682c1bb4e523}


Clear out all of the Genomic\-Region objects \index{GenomicRegionVector@{Genomic\-Region\-Vector}!concat@{concat}}
\index{concat@{concat}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Genomic\-Region\-Vector::concat (const \bf{Genomic\-Region\-Vector} \& {\em g})}\label{classGenomicRegionVector_894ce6c8a5b366f5fb3f75a310d2499e}


Add two \doxyref{Genomic\-Region\-Vector}{p.}{classGenomicRegionVector} objects together \index{GenomicRegionVector@{Genomic\-Region\-Vector}!createTreeMap@{createTreeMap}}
\index{createTreeMap@{createTreeMap}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Genomic\-Region\-Vector::create\-Tree\-Map ()}\label{classGenomicRegionVector_45fd5c44e0851148b17146318795404d}


Fill in the Genomic\-Interval\-Tree\-Map stored in this object.

A Genomic\-Interval\-Tree\-Map is an unordered\_\-map of Genomic\-Interval\-Trees for each chromosome. A Genomic\-Interval\-Tree is an interval tree on the ranges defined by the genomic interval, with cargo set at the same Genomic\-Region object. \index{GenomicRegionVector@{Genomic\-Region\-Vector}!empty@{empty}}
\index{empty@{empty}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}bool Genomic\-Region\-Vector::empty () const\hspace{0.3cm}{\tt  [inline]}}\label{classGenomicRegionVector_aea0228989aeb752d857b213eeaf8e2d}


Is this object empty? \index{GenomicRegionVector@{Genomic\-Region\-Vector}!findOverlapping@{findOverlapping}}
\index{findOverlapping@{findOverlapping}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}size\_\-t Genomic\-Region\-Vector::find\-Overlapping (const Genomic\-Region \& {\em gr})}\label{classGenomicRegionVector_b9895bc9c23c687561b3c23ac0f23320}


Find overlaps between this vector and input Genomic\-Region.

Requires that the Genomic\-Interval\-Tree\-Map have been created first \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em gr}]Region to test \end{description}
\end{Desc}
\begin{Desc}
\item[Returns:]Number of overlapping elements in this \doxyref{Genomic\-Region\-Vector}{p.}{classGenomicRegionVector} \end{Desc}
\index{GenomicRegionVector@{Genomic\-Region\-Vector}!getNextGenomicRegion@{getNextGenomicRegion}}
\index{getNextGenomicRegion@{getNextGenomicRegion}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}bool Genomic\-Region\-Vector::get\-Next\-Genomic\-Region (Genomic\-Region \& {\em gr})}\label{classGenomicRegionVector_9880c8cdbe9ac50d8a579b483d0649f5}


Fill the next Genomic\-Region object. \begin{Desc}
\item[Returns:]false if add end of vector \end{Desc}
\index{GenomicRegionVector@{Genomic\-Region\-Vector}!mergeOverlappingIntervals@{mergeOverlappingIntervals}}
\index{mergeOverlappingIntervals@{mergeOverlappingIntervals}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Genomic\-Region\-Vector::merge\-Overlapping\-Intervals ()}\label{classGenomicRegionVector_a30d9f4def4c7ac7958a4475f2997df5}


Reduces the Genomic\-Region objects to minimal set \index{GenomicRegionVector@{Genomic\-Region\-Vector}!readBEDfile@{readBEDfile}}
\index{readBEDfile@{readBEDfile}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Genomic\-Region\-Vector::read\-BEDfile (const std::string \& {\em file}, int {\em pad} = {\tt 0})}\label{classGenomicRegionVector_ba5589c933030ffb64dbeaf8350276b2}


Read in a BED file and adds to \doxyref{Genomic\-Region\-Vector}{p.}{classGenomicRegionVector} object \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em file}]Path to BED file \item[{\em pad}]Amount to pad intervals by \end{description}
\end{Desc}
\index{GenomicRegionVector@{Genomic\-Region\-Vector}!readMuTect@{readMuTect}}
\index{readMuTect@{readMuTect}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Genomic\-Region\-Vector::read\-Mu\-Tect (const std::string \& {\em file}, int {\em pad} = {\tt 0})}\label{classGenomicRegionVector_e4a4d90b656a928773581de39a26c580}


Read in a Mu\-Tect call-stats file and adds to \doxyref{Genomic\-Region\-Vector}{p.}{classGenomicRegionVector} object.

Reads a Mu\-Tect call-stats file and imports only lines with KEEP marked. \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em file}]Path to call-stats file \item[{\em pad}]Amount to pad intervals by \end{description}
\end{Desc}
\index{GenomicRegionVector@{Genomic\-Region\-Vector}!readVCFfile@{readVCFfile}}
\index{readVCFfile@{readVCFfile}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Genomic\-Region\-Vector::read\-VCFfile (const std::string \& {\em file}, int {\em pad} = {\tt 0})}\label{classGenomicRegionVector_27b8fa36a7b5b81f5abe2dad8dd752a7}


Read in a VCF file and adds to \doxyref{Genomic\-Region\-Vector}{p.}{classGenomicRegionVector} object \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em file}]Path to BED file \item[{\em pad}]Amount to pad intervals by \end{description}
\end{Desc}
\index{GenomicRegionVector@{Genomic\-Region\-Vector}!regionFileToGRV@{regionFileToGRV}}
\index{regionFileToGRV@{regionFileToGRV}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Genomic\-Region\-Vector::region\-File\-To\-GRV (const std::string \& {\em file}, int {\em pad} = {\tt 0})}\label{classGenomicRegionVector_e3e69a79c878ae9d4f62b74e5f08bc5b}


Read in a text file (can be gzipped) and add to \doxyref{Genomic\-Region\-Vector}{p.}{classGenomicRegionVector}

This function will automatically detect which file type is being input: -- ends in .vcf -$>$ read\-VCFfile -- ends in .bed -$>$ read\-BEDfile -- header contains \char`\"{}Mu\-Tect\char`\"{} -$>$ read\-Mu\-Tect The values are appended to existing vector of Genomic\-Region objects \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em file}]Text file to read and store intervals \item[{\em pad}]Amount to pad the intervals by (calls Genomic\-Region::pad) \end{description}
\end{Desc}
\index{GenomicRegionVector@{Genomic\-Region\-Vector}!rewind@{rewind}}
\index{rewind@{rewind}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Genomic\-Region\-Vector::rewind ()\hspace{0.3cm}{\tt  [inline]}}\label{classGenomicRegionVector_134e5edbdb66be9149dcc1bae97c4b7a}


Rewind the element pointer to the first Genomic\-Region \index{GenomicRegionVector@{Genomic\-Region\-Vector}!sendToBED@{sendToBED}}
\index{sendToBED@{sendToBED}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}string Genomic\-Region\-Vector::send\-To\-BED () const}\label{classGenomicRegionVector_2fddfd8ce536ef6ea82f221a1fbaa9a1}


Output the \doxyref{Genomic\-Region\-Vector}{p.}{classGenomicRegionVector} to a BED format

Currently just outputs chr pos pos2 with no header. \begin{Desc}
\item[Returns:]BED formated string reprsentation \end{Desc}
\index{GenomicRegionVector@{Genomic\-Region\-Vector}!sendToBED@{sendToBED}}
\index{sendToBED@{sendToBED}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Genomic\-Region\-Vector::send\-To\-BED (const std::string {\em file})}\label{classGenomicRegionVector_d6c335703ad494ce93e0e87e8bf9e493}


Send the \doxyref{Genomic\-Region\-Vector}{p.}{classGenomicRegionVector} to a BED file \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em file}]Output file \end{description}
\end{Desc}
\index{GenomicRegionVector@{Genomic\-Region\-Vector}!size@{size}}
\index{size@{size}!GenomicRegionVector@{Genomic\-Region\-Vector}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}size\_\-t Genomic\-Region\-Vector::size () const\hspace{0.3cm}{\tt  [inline]}}\label{classGenomicRegionVector_7eefd44e17d6a72a252ce733381982b3}


Return the number of Genomic\-Regions stored 

The documentation for this class was generated from the following file:\begin{CompactItemize}
\item 
/xchip/gistic/Jeremiah/GIT/isva/Snow\-Tools/src/Snow\-Tools/Genomic\-Region\-Vector.h\end{CompactItemize}
